File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Reverse And Merge
# O(n) time, O(1) space
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # Find middle
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Reverse the second portion
        second = slow.next
        prev = slow.next = None
        while second:
            next = second.next
            second.next = prev
            prev = second
            second = next
        
        # Merge two halves
        first, second = head, prev
        while second:
            temp1, temp2 = first.next, second. next
            first.next = second
            second.next = temp1
            first, second = temp1, temp2

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    private fun mergeList(l1: ListNode?, l2: ListNode?): ListNode? {
        val dummy = ListNode(0)
        var tail = dummy
        var first = l1
        var second = l2

        while (first != null && second != null) {
            if (first.`val` < second.`val`) {
                tail.next = first
                first = first.next
            } else {
                tail.next = second
                second = second.next
            }
            tail = tail.next!!
        }

        if (first != null) {
            tail.next = first
        }
        if (second != null) {
            tail.next = second
        }

        return dummy.next
    }

    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        if (lists.isEmpty()) {
            return null
        }

        for (i in 1 until lists.size) {
            lists[i] = mergeList(lists[i-1], lists[i])
        }

        return lists.last()
    }
}